feature/support for NGSIv2

cblanco opened the pull request.

cblanco commented on documentation/configuration.md:
```diff
@@ -60,7 +60,7 @@ In order to have perseo running, there are several basic pieces of information t
 * `config.smpp.from`: Number from SMS are sending by SMPP server
 * `config.smpp.enabled`: SMPP is default method for SMS instead of use SMS gateway.
-* `config.orion.URL`: URL for updating contexts at Orion (Context Broker).
+* `config.orion.URL`: Context Broker URL, e.g. https://orion.example.com
```
This PR, needs a change in the configuration that we need to validate.

The URL for the updateContext can not include the path, since it is not the same for NGSIv2. So we propose to change this variable by the URL of the context Broker and add the path depending on whether the update is v1 or v2.

fgalan replied:
After internal discussion we think is a good idea to redefine the semantic of `config.orion.URL` (and associated PERSEO_ORION_URL env var) to be the base URL of Orion (so it gets complented by `/v1/updateContext` in the case of NGSIv1 update action and `/v2/entities?options=upsert` in the case of NGSIv2 update action).

However, I'd suggest to use "base URL" intead of "URL" (in can be confussing) and use an example closer to a real Orion (i.e. using port 1026). In sum:

```
`config.orion.URL`: Context Broker base URL, e.g. https://orion.example.com:1026
```

cblanco replied:
Fixed in d30f1ed

cblanco commented on lib/models/updateAction.js:
```diff
@@ -276,9 +276,9 @@ function doRequestV2(action, event, callback) {
-        // TODO token (fiware & OpenStack)
-    }
+    // if (action.parameters.trust) {
+    //     // TODO token (fiware & OpenStack)
```
We need to talk about how to proceed with the issue of tokens.

fgalan commented:
We have had this morning a fruitful review session at Google Handout. This is my summary in the case may be useful:

* Try to add JavaDoc in new methods (not a must, but very convenient)
* Trust token aspects in NGSIv2 update actions are deferred to a next PR. This PR leaves a TODO to remark the missing parts.
* The heuristic used to identify NGSIv2 notifications should be explained in a brief comment at https://github.com/telefonicaid/perseo-fe/pull/305/files#diff-ed9fd8f8dc7389ca8da8cdaa8eaa65d1R325
* A typo (DateType) detected at https://github.com/telefonicaid/perseo-fe/pull/305/files#diff-43c1a406c68dbf35ee797980d6caf5deR257
* Avoid using  `^` in dependences. Use exact version or `~`
* Both `"2"` and `2` should be valid
* Missing license header in some files.
* We (@fgalan @mrutid ) will have a closer look to configuration changes involved in this PR and provide feedback as soon as possible.

And an additional one that I have realized after the meeting (sorry...)

* This PR doesn't modify any .md file so I understand that documenation has not been modified. At least the documentation about the update action should be modified to explain the new "version" optional field. Maybe another documents needs also to be adjusted.

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -10,3 +10,6 @@
   * closure-linter-wrapper
+- Change on the PERSEO_ORION_URL env var behaviour. Now it represents Context Broker URL instead of the updateContext
+  endpoint
+- Add: NGSIv2 support
```
I'd suggest to be more precise here. I mean, something like this:

```
Add: NGSIv2 support in both notification reception and CB update action
```

In addition, I'd suggest to mention the added dependenciy (ngsijs) and remove chai, sinon and sinon-chai from the list of removed dependencies ;) In fact, if the re-introduced version is different from the old one a line line this one "Upgrade dev dependency chai|sinon|sinon-chai from ... to ..." should be added.

cblanco replied:
Fixed in 330b1b0

fgalan commented:
> We (@fgalan @mrutid ) will have a closer look to configuration changes involved in this PR and provide feedback as soon as possible.

See https://github.com/telefonicaid/perseo-fe/pull/305/files#r238354925

cblanco commented:
This is a big PR. We are solving the problems little by little:
> Try to add JavaDoc in new methods (not a must, but very convenient)
Fixed in dcbb3ea
> Trust token aspects in NGSIv2 update actions are deferred to a next PR. This PR leaves a TODO to remark the missing parts.
Fixed in b865282
> The heuristic used to identify NGSIv2 notifications should be explained in a brief comment at https://github.com/telefonicaid/perseo-fe/pull/305/files#diff-ed9fd8f8dc7389ca8da8cdaa8eaa65d1R325
Fixed in 70631be
> A typo (DateType) detected at https://github.com/telefonicaid/perseo-fe/pull/305/files#diff-43c1a406c68dbf35ee797980d6caf5deR257
Fixed in 2044f95
> Avoid using ^ in dependences. Use exact version or ~
Fixed in 9a79c29
> Both "2" and 2 should be valid
Fixed in 322d3f1
> Missing license header in some files.
Fixed in 5065a58
> This PR doesn't modify any .md file so I understand that documenation has not been modified. At least the documentation about the update action should be modified to explain the new "version" optional field. Maybe another documents needs also to be adjusted.
Fixed in 
3ab3f09

cblanco commented:
The coverage of the tests has been reduced in the update_actions file due to the last changes introduced in the processOptionParams method. I will try to add unit tests to cover this code soon. I think that I will need to use the rewire library to test the private methods.

It would be interesting to put cases of update actions NGSIv2 in the acceptance tests that are included in the repository. I can not execute them on my own because I think they depend on private repositories that we do not have access to.

fgalan commented:
> It would be interesting to put cases of update actions NGSIv2 in the acceptance tests that are included in the repository. I can not execute them on my own because I think they depend on private repositories that we do not have access to.

I understand you mean Python-behave tests that are included in this repository. They are mostly outdated, so don't pay too much attention to it. At the present moment, only unit tests should be taken into account.

fgalan commented on documentation/plain_rules.md:
```diff
@@ -190,6 +191,50 @@ First time an update action using trust token is triggered, Perseo interacts wit
+
+**Note:** NGSIv2 update actions ignore the trust token for now.
+
+When using NGSIv2 in the update actions, the value field perform [string substitution](#string-substitution-syntax). If `value` is a String, Perseo will parse the value taking into account the `type` field, this only applies to *`Number`*, *`Boolean`* and *`None`* types. 
```
This is a bit obsure... What do you think about adding an example to show how it works?

cblanco replied:
Yes. I have reviewed the test cases and I have realized that it was necessary to parse the DateTimes correctly in order to directly use rules variables without using cast. I added the case in this commit 5ccccb7. An more tests for all possible types in 2c7902c.

New documentation for NGSIv2 types in this commit 3d148d6.

fgalan commented on package.json:
```diff
@@ -32,9 +32,13 @@
     "mocha": "5.2.0",
+    "chai": "~4.1.2",
     "proxyquire": "0.5.1",
+    "rewire": "^4.0.1",
```
Please use ~ instead of ^ and add to CHANGES_NEXT_RELEASE file.

cblanco replied:
Fixed in 71ea6cc

fgalan commented on test/unit/notices_processCBNotice.js:
```diff
@@ -0,0 +1,452 @@
+                });
+            }
+        );
+        // why NGSIv1 dont parse metadata geo:points? only parseocation when 'name' is 'location'.
```
This is a real question? :)

Answer: `geo:...` attribute types are part of the NGSIv2 specification. It's usage in NGSIv1 is CB is undefied (not sure right now... maybe there is some degree of support, but I don't remember pretty well right now).

fgalan replied:
This block of code should be removed? We don't like to leave dead code behind us, do we? ;)

cblanco replied:
Ooops jejeje. This is not a real question, it's just my internal questions. I do not have much experience with NGSIv1 :)

Fixed in 24ade02

cblanco commented on test/unit/updateAction.js:
```diff
@@ -0,0 +1,283 @@
+'use strict';
```
Fixed in 9d39257

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -5,8 +5,13 @@
+- Change on the PERSEO_ORION_URL env var behaviour. Now it represents Context Broker base URL instead of the
+  updateContext endpoint
+- Add: new ngsijs ~1.2.0 dependency
+- Add: new rewire ~~.0.1 dependency
```
I guess that ~~ is a typo...

cblanco replied:
Fixed in 775442c

fgalan commented on documentation/plain_rules.md:
````diff
@@ -190,6 +191,197 @@ First time an update action using trust token is triggered, Perseo interacts wit
+}
+```
+
+Note that using NGSIv2 the BloodPressure attribute is a Number and therefore it is not necessary to use `cast()`.
````
I like this! :)

Current rules do an overuse `cast()` and NGSIv2 will allow to simplify a lot these EPL framents.

NTC (just a comment)

fgalan commented on documentation/plain_rules.md:
```diff
@@ -190,6 +191,197 @@ First time an update action using trust token is triggered, Perseo interacts wit
+            "attributes": [
+                {
+                    "name":"abnormal",
+                    "type":"number",
```
Maybe this examples should change to "Number" (with capital letter)? Or the type processing logic is case insensitive?

fgalan replied:
Looking to switch() at https://github.com/telefonicaid/perseo-fe/pull/305/files#diff-43c1a406c68dbf35ee797980d6caf5deR267 it seems case sensitive.

cblanco replied:
Yes. is case sensitive. This "number" in the type is an error.

Fixed in 3e85263

cblanco commented on CHANGES_NEXT_RELEASE:
```diff
@@ -5,8 +5,13 @@
+- Add: new ngsijs ~1.2.0 dependency
+- Add: new rewire ~~.0.1 dependency
+- Add: NGSIv2 support in both notification reception and CB update action
+- Upgrade dev dependency chai from ~1.8.0" to ~4.1.2
```
Fixed in 775442c

cblanco commented on documentation/plain_rules.md:
````diff
@@ -190,6 +191,197 @@ First time an update action using trust token is triggered, Perseo interacts wit
+- Number from variable:
+```json
+{
+    'name':'numberFromValue',
````
Fixed in 8ceb8bf

cblanco commented:
Typo in CNR Fixed in ab40223

Thank you very much for the review @fgalan. It is a pleasure to collaborate in the development of this component. :)
